LFLT-523 Implement controller for file handling in LSRS

petersmith-hun opened the pull request:
- Implemented file operations controllers
 - Implemented support logic for validation, input mapping, error handling and file upload
 - Bumped application version to 0.5.0-dev
 - Implemented unit tests

petersmith-hun committed:
LFLT-523 Implement controller for file handling in LSRS
 - Implemented file operations controllers
 - Implemented support logic for validation, input mapping, error handling and file upload
 - Bumped application version to 0.5.0-dev

petersmith-hun committed:
LFLT-523 Implement controller for file handling in LSRS
 - Implemented unit tests

elementbound commented on src/lsrs/web/controller/actuator-controller.ts:
```diff
@@ -25,7 +25,7 @@ export default class ActuatorController implements Controller {
      * @returns application info as InfoResponse wrapped in ResponseWrapper
      */
-    public info(): ResponseWrapper<InfoResponse> {
+    public async info(): Promise<ResponseWrapper<InfoResponse>> {
```
Why is this async?

petersmith-hun replied:
To keep it consistent and out of pure laziness as well. :) But your other question, regarding the MappingHelper could help here. I'll actually take a look at it. If I can get rid of the Promise there, I could revert this, too.

petersmith-hun replied:
Your suggestion regarding the MappingHelper#register methods made it possible to revert this change, so thanks! :)

elementbound commented on src/lsrs/web/converter/converters.ts:
```diff
@@ -0,0 +1,69 @@
+ */
+export function convertAcceptor(acceptor: AcceptorInfo): DirectoryModel {
+
+    return {
```
Would it make sense to simply cast the acceptor and return that? Since the properties seem to match up.

petersmith-hun replied:
Not exactly, but maybe I could that.. Will check!

petersmith-hun replied:
Yep, it did make sense, dropped the converter and modified the AcceptorInfo internal domain class. Thanks for the suggestion!

elementbound commented on src/lsrs/web/model/common.ts:
```diff
@@ -13,10 +31,50 @@ export enum HttpStatus {
 
     readonly status: HttpStatus;
-    readonly content: T;
+    readonly content: Optional<T>;
```
Why not a nullable property?
`readonly content?: T`

petersmith-hun replied:
Because that did not work with "null" (null vs undefined).

elementbound replied:
Why would you use null?

petersmith-hun replied:
You won't be happy about my answer. :D

petersmith-hun replied:
Changed to `content?: T`, it also let me simplify the response mapping logic, thanks!

elementbound commented on src/lsrs/web/utility/mapping-helper.ts:
```diff
@@ -0,0 +1,101 @@
+ * controllers might expect a specific request model, this helper component also deals with converting the "raw" request
+ * object into the defined request model. Also attaches a response handler and an exception handler to each of the registrations.
+ */
+abstract class MappingHelper {
```
Nice!

petersmith-hun replied:
Thanks. :3

elementbound commented on src/lsrs/web/utility/mapping-helper.ts:
```diff
@@ -0,0 +1,101 @@
+
+        response.status(responseWrapper.status);
+
+        if (responseWrapper.content) {
```
Depending on how concerned you are with sending `0` or `""` as response body, you could also explicitly check for `content !== undefined`

petersmith-hun replied:
Currently it's always either set to null or there's something in the content field (other than a 0 or an empty string). A simple "is there anything in it?" check (like currently) should be enough.

elementbound commented on src/lsrs/web/utility/mapping-helper.ts:
```diff
@@ -0,0 +1,101 @@
+        response.status(responseWrapper.status);
+
+        if (responseWrapper.content) {
+            if (responseWrapper.sendAsRaw) {
```
If you're in for the zoomer points, you could also use a convenient ternary operator expression:
```
responseWrapper.sendAsRaw
    ? response.send(responseWrapper.content)
    : response.json(responseWrapper.content) 
```

petersmith-hun replied:
Awesome note, and I wanted to do that actually, but could not. If the content field is null, .send() is sending a "null" value in the response body. I didn't try it with undefined, but I guess the something similar would happen (I mean, "undefined" in the response body)..? If not, then I'm up to some minor refactor... :D

elementbound replied:
Yep, I only meant the inner part, including the outer would start to get confusing:
```
responseWrapper.content
    ? (responseWrapper.sendAsRaw
        ? response.send(responseWrapper.content)
        : response.json(responseWrapper.content))
    : response.send()
```

petersmith-hun replied:
I'd rather not start embedding ternary conditional expressions, that's pretty confusing, yeah. :/

petersmith-hun replied:
And didn't actually need to, I managed to simplify the logic a bit, after changing the content type to be optional.

elementbound commented on src/lsrs/web/utility/mapping-helper.ts:
```diff
@@ -0,0 +1,101 @@
+     * @param endpointCall controller endpoint call to be registered
+     * @returns Express request handler definition
+     */
+    public register(endpointCall: () => Promise<ResponseWrapper<any>>): (request: Request, response: Response, next: NextFunction) => void {
```
Would it make sense to spec the parameter lambda return type as `Promise<ResponseWrapper<any>> | ResponseWrapper<any>`? I think you can await it either way, and that way you wouldn't force using async methods even if they don't need to await anything.

petersmith-hun replied:
Recursive reference back to your first comment. :) Will check it, but I believe it was raising a syntax error, when I didn't define the Promise<> generic type in the specification.

petersmith-hun replied:
Yes, it totally made sense, thanks! It also let me revert the actuator endpoint changes.

petersmith-hun committed:
LFLT-523 Implement controller for file handling in LSRS
 - Code review comments
 - Modified AcceptorInfo and removed the relevant converter to DirectoryModel
 - Reverted making the actuator endpoints async
 - Changed content field in ResponseWrapper to optional instead of being nullable
 - Response content related refactor in MappingHelper
 - Changed register method of MappingHelper implementations to accept sync endpoint calls as well
